junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,474 @@
+ *                annotations = List(new Test, new Ignore("baz not implemented yet")))
+ *        ),
+ *        moduleMethods(
+ *            new JUnitMethodMetadata(name = "beforeClass",
+ *                annotations = List(new BeforeClass)))
+ *      )
+ *    }
+ *
+ *    def newInstance(): AnyRef = new Foo()
+ *
+ *    def invoke(methodName: String): Unit = {
+ *      if (methodName == "0") Foo.beforeClass()
+ *      else throw new NoSuchMethodException(methodId)
+ *    }
+ *
+ *    def invoke(instance: AnyRef, methodName: String): Unit = {
+ *      if (methodName == "before") instance.asInstanceOf[Foo].before()
+ *      else if (methodName == "bar") instance.asInstanceOf[Foo].bar()
+ *      else if (methodName == "baz") instance.asInstanceOf[Foo].baz()
+ *      else throw new NoSuchMethodException(methodId)
+ *    }
+ *  }
+ *  }}}
+ *  The test framework will identify `Foo\$scalajs\$junit\$bootstrapper` as a test module
+ *  because it extends `JUnitTestBootstrapper`. It will know which methods to run based
+ *  on the info returned by Foo\$scalajs\$junit\$bootstrapper.metadata,
+ *  it will create new test instances using `Foo\$scalajs\$junit\$bootstrapper.newInstance()`
+ *  and it will invoke test methods using `invoke` on the bootstrapper.
+ */
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Makes JUnit test classes invokable in Scala.js"
+
+  /* Hack to be able to access `registerModuleExports` of ScalaJSPlugin
+   * without modifying it.
+   */
+  private lazy val scalaJSPlugin = {
+    type ScalaJSPlugin = NscPlugin {
+      def registerModuleExports(sym: ScalaJSJUnitPluginComponent.global.Symbol): Unit
+    }
+    global.plugins.collectFirst {
+      case pl if pl.getClass.getName == "org.scalajs.core.compiler.ScalaJSPlugin" =>
+        pl.asInstanceOf[ScalaJSPlugin]
+    }.getOrElse {
+      throw new Exception(
+          "The Scala.js JUnit plugin only works with the Scala.js plugin enabled.")
+    }
+  }
+
+  object ScalaJSJUnitPluginComponent
+      extends plugins.PluginComponent with transform.Transform with Compat210Component {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = ScalaJSJUnitPlugin.this.global
+
+    import global._
+
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends Transformer {
+
+      import rootMirror.getRequiredClass
+
+      private val testAnnotationSymbol =
+        getRequiredClass("org.junit.Test")
+
+      private val fixMethodOrderAnnotationSymbol =
+        getRequiredClass("org.junit.FixMethodOrder")
```

Comment:
I think (unwritten) convention in Scala and Scala.js is to call these things `XYZClass`.